controller: bml: Revive bml_set_wifi_credentials (part 1/2)

VladyslavTupikin opened the pull request:
bml_set_wifi_credentials is an old implementation for a CLI
command for changing credentials for a specific radio. It needs
to be revived to allow setting credentials for any agent in the
network. Detailed description was given in the issue # 1058, but
without UML diagrams for current implementation and proposed
functions like bml_set_wifi_credentials, bml_clear_wifi_credentials,
bml_update_wifi_credentials.

Add UML diagrams for each new feature and current implementation.

The implementation of bml_set_wifi_credentials itself is for a second PR.

VladyslavTupikin commented:
UML diagram for current bml_set_wifi_credentials implementation:
![image](https://user-images.githubusercontent.com/61505283/78826979-0c5e3f00-79eb-11ea-80cf-d86d60d1fdb8.png)

UML diagram for new bml_set_wifi_credentials implementation:
![image](https://user-images.githubusercontent.com/61505283/78984380-c3170800-7b2e-11ea-8704-1f6fc081de07.png)

UML diagram for new bml_clear_wifi_credentials implementation:
![image](https://user-images.githubusercontent.com/61505283/78984455-f194e300-7b2e-11ea-82eb-57b96410640a.png)

UML diagram for new bml_update_wifi_credentials implementation:
![image](https://user-images.githubusercontent.com/61505283/78984538-2b65e980-7b2f-11ea-9a11-e4d9a69e80f0.png)

arnout commented on documentation/images/plantuml/beerocks_cli/clear_credantionals.puml:
```diff
@@ -0,0 +1,10 @@
+@startuml
```
File should be named `clear_credentials.puml`

Also, the commit message summary line is not OK. It should describe what that commit does, and the body should describe the why. So:

```
documentation: add UML files for BML credentials commands
```

The body of the commit message is excellent.

arnout commented on documentation/images/plantuml/beerocks_cli/clear_credantionals.puml:
```diff
@@ -0,0 +1,10 @@
+@startuml
+ -> beerocks_cli: beerocks_cli -c \n"bml_clear_wifi_credentials al-mac"
```
```suggestion
 -> beerocks_cli: beerocks_cli -c \n"bml_clear_wifi_credentials <al-mac>"
```
(nitpick)

arnout commented on documentation/images/plantuml/beerocks_cli/clear_credantionals.puml:
```diff
@@ -0,0 +1,10 @@
+bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST)
+controller -> controller_db: clear_bss_info_configuration(\n const sMacAddr &al_mac)
+controller_db -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE)
+@enduml
```
Please make sure there's a newline at the end of the file.

arnout commented on documentation/images/plantuml/beerocks_cli/set_credentials_old.puml:
```diff
@@ -0,0 +1,9 @@
+beerocks_cli -> beerocks_cli: bml_set_wifi_credentials(\n ctx, \nssid.c_str(), \npass.c_str(), \nsec, \nvap_id, \nforce)
+beerocks_cli -> bml_internal: set_wifi_credentials(\n const std::string ssid,\n const std::string pass,\n int sec,\n int vap_id,\n int force)
+bml_internal -> controller: message_com::send_cmdu(m_sockMaster, ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST)
+note right: // TODO: trigger auto-config, probaly\n it will be a good idea to change the message\n to "ACTION_BML_UPDATE_CONFIGURATION_REQUEST"
```
We don't actually need the old one in the repository. So maybe a better way would be to split this into two commits: the first commit just adds `set_credentials.puml` that describes the current situation (i.e., set_credentials_old), and the second commit update the same file to the new situation (i.e., set_credentials_new). The diff then nicely shows what changes from old to new.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -392,6 +392,15 @@ void cli_bml::setFunctionsMapAndArray()
         static_cast<pFunction>(&cli_bml::set_wifi_credentials_caller), 2, 5, STRING_ARG, STRING_ARG,
         INT_ARG, INT_ARG, INT_ARG);
+    insertCommandToMap(
+        "bml_set_wifi_credentials_new", "<AL-MAC> <ssid> <network_key> <radio> <bss_type>",
```
Don't bother with a new one. I understand that you want to avoid breaking stuff. However, at some point, we will anyway have to remove the old one, so we can just as well do it now.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -392,6 +392,15 @@ void cli_bml::setFunctionsMapAndArray()
         INT_ARG, INT_ARG, INT_ARG);
+    insertCommandToMap(
+        "bml_set_wifi_credentials_new", "<AL-MAC> <ssid> <network_key> <radio> <bss_type>",
+        "Sets new wifi credentials for the given AL-MAC."
```
It's not actually setting them, it's adding them.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -392,6 +392,15 @@ void cli_bml::setFunctionsMapAndArray()
+        "Set network_key, (passphrase) "
+        "Set radio type 24g or 5g "
+        "Set bss_type, can be fronthaul, backhaul, fronthaul-backhaul",
+        static_cast<pFunction>(&cli_bml::set_wifi_credentials_new_caller), 2, 5, STRING_ARG,
```
```suggestion
        static_cast<pFunction>(&cli_bml::set_wifi_credentials_new_caller), 5, 5, STRING_ARG,
```
It's always 5 arguments, right? None are optional?

Although it would actually be good to default radio type to 24g-5g and bss type to fronthaul.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -392,6 +392,15 @@ void cli_bml::setFunctionsMapAndArray()
+        "Sets new wifi credentials for the given AL-MAC."
+        "Set new ssid "
+        "Set network_key, (passphrase) "
+        "Set radio type 24g or 5g "
```
Maybe also allow 24g-5g for both.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -932,6 +941,23 @@ int cli_bml::set_wifi_credentials_caller(int numOfArgs)
+        return -1;
+}
+
+int cli_bml::clear_wifi_credentials_caller(int numOfArgs)
+{
+    if (numOfArgs == 1)
+        return clear_wifi_credentials(args.stringArgs[0]);
+    else
+        return -1;
+}
```
There should be a single commit that does all of the clear functionality.

arnout commented on documentation/images/plantuml/beerocks_cli/set_credentials_new.puml:
```diff
@@ -0,0 +1,10 @@
+beerocks_cli -> beerocks_cli: cli_bml::set_wifi_credentials(\n args.stringArgs[0],\n args.stringArgs[1],\n args.stringArgs[2],\n args.stringArgs[3],\n args.stringArgs[4])
+beerocks_cli -> beerocks_cli: bml_set_wifi_credentials(\n ctx,\n al-mac.c_str(),\n ssid.c_str(),\n network_key.c_str(),\n radio.c_str(),\n bss_type.c_str())
+beerocks_cli -> bml_internal: set_wifi_credentials(\n const son::wireless_utils::sBssInfoConf &credentials,\n const sMacAddr &al_mac)
+bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST)
```
```suggestion
bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_ADD_REQUEST)

```

arnout commented on documentation/images/plantuml/beerocks_cli/set_credentials_new.puml:
```diff
@@ -0,0 +1,10 @@
+beerocks_cli -> bml_internal: set_wifi_credentials(\n const son::wireless_utils::sBssInfoConf &credentials,\n const sMacAddr &al_mac)
+bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST)
+controller -> controller_db: add_bss_info_configuration(\n const sMacAddr &al_mac,\n const wireless_utils::sBssInfoConf &bss_info)
+controller_db -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE)
```
```suggestion
controller_db -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_ADD_RESPONSE)

```

arnout commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -151,6 +151,46 @@ int bml_event_register_cb(BML_CTX ctx, BML_EVENT_CB cb);
+ *
+ * @return BML_RET_OK on success.
+ */
+int bml_set_wifi_credentials_new(BML_CTX ctx,const char* al_mac, const char* ssid, const char* network_key,
```
Make sure you run clang-format before committing.

arnout commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -151,6 +151,46 @@ int bml_event_register_cb(BML_CTX ctx, BML_EVENT_CB cb);
+ *
+ * @return BML_RET_OK on success.
+ */
+int bml_update_wifi_credentials(BML_CTX ctx,const char* al_mac, const char* ssid, const char* network_key,
```
"update" just commits the credentials that were added before with set_wifi_credentials. So update takes only al_mac as argument.

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1944,6 +1944,30 @@ int bml_internal::register_event_cb(BML_EVENT_CB pCB)
+    return BML_RET_OK;
+}
+
+int bml_internal::update_wifi_credentials(const son::wireless_utils::sBssInfoConf &credentials,
```
Again, update only takes al_mac as argument.

arnout commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -55,6 +56,40 @@ class bml_internal : public beerocks::socket_thread {
+    /**
+     * @brief update wifi credentials for specific AL-MAC
+     * 
+     * Sets wifi credentials for specific al_mac and sends autoconfig renew if requested.
```
```suggestion
     * Sends autoconfig renew if credentials have changed for AL-MAC.
```

Note that for now, we'll simply *always* send autoconfig renew, even if credentials haven't changed. We can add a TODO somewhere to fix that.

arnout commented on documentation/images/plantuml/beerocks_cli/clear_credantionals.puml:
```diff
@@ -0,0 +1,10 @@
+@startuml
+ -> beerocks_cli: beerocks_cli -c \n"bml_clear_wifi_credentials al-mac"
+beerocks_cli -> beerocks_cli: cli_bml::clear_wifi_credentials_caller(args)
+beerocks_cli -> beerocks_cli: cli_bml::clear_wifi_credentials(\n args.stringArgs[0],\n args.stringArgs[1],\n args.stringArgs[2],\n args.stringArgs[3],\n args.stringArgs[4])
```
```suggestion
beerocks_cli -> beerocks_cli: cli_bml::clear_wifi_credentials(\n args.stringArgs[0])

```

arnout commented on documentation/images/plantuml/beerocks_cli/update_credentials.puml:
```diff
@@ -0,0 +1,10 @@
+beerocks_cli -> beerocks_cli: cli_bml::update_wifi_credentials_caller(args)
+beerocks_cli -> beerocks_cli: cli_bml::update_wifi_credentials(\n args.stringArgs[0],\n args.stringArgs[1],\n args.stringArgs[2],\n args.stringArgs[3],\n args.stringArgs[4])
+beerocks_cli -> beerocks_cli: bml_update_wifi_credentials(\n ctx,\n al-mac.c_str(),\n ssid.c_str(),\n network_key.c_str(),\n radio.c_str(),\n bss_type.c_str())
+beerocks_cli -> bml_internal: update_wifi_credentials(\n const son::wireless_utils::sBssInfoConf &credentials,\n const sMacAddr &al_mac)
```
```suggestion
beerocks_cli -> beerocks_cli: cli_bml::update_wifi_credentials(\n args.stringArgs[0])
beerocks_cli -> beerocks_cli: bml_update_wifi_credentials(\n ctx,\n al-mac.c_str())
beerocks_cli -> bml_internal: update_wifi_credentials(\n const sMacAddr &al_mac)
```

arnout commented on documentation/images/plantuml/beerocks_cli/update_credentials.puml:
```diff
@@ -0,0 +1,10 @@
+beerocks_cli -> beerocks_cli: bml_update_wifi_credentials(\n ctx,\n al-mac.c_str(),\n ssid.c_str(),\n network_key.c_str(),\n radio.c_str(),\n bss_type.c_str())
+beerocks_cli -> bml_internal: update_wifi_credentials(\n const son::wireless_utils::sBssInfoConf &credentials,\n const sMacAddr &al_mac)
+bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST)
+controller -> controller_db: update_bss_info_configuration(\n const sMacAddr &al_mac,\n const wireless_utils::sBssInfoConf &bss_info)
```
This doesn't do anything in the controller db. Instead, it sends out an AUTOCONFIG_RENEW message over the network. So this should rather be something like:
```suggestion
controller -> controller: send_cmdu_to_bus(\n AP_AUTOCONFIGURATION_RENEW_MESSAGE,\n al_mac, db.get_local_bridge_mac());
```

Commit:
documentation: add UML file for BML set credential command

Current implementation for bml_set_wifi_credentials does
not have UML diagram with explanations of this feature.

Add uml diagram which explains current situation for the main
APIs of this feature.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
documentation: add UML files for BML credentials commands

bml_set_wifi_credentials is an old implementation for a CLI
command for changing credentials for a specific radio. It needs
to be revived to allow setting credentials for any agent in the
network. Detailed description was given in the issue # 1058, but
without UML diagrams for proposed functions like
bml_set_wifi_credentials, bml_clear_wifi_credentials,
bml_update_wifi_credentials.

Add UML diagrams for each new feature.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
beerocks: bcl: Add 2 methods for conversion

Need to create 2 methods: first for conversion string to
the operating class list, second for conversion string
to the WSC::eWscVendorExtSubelementBssType.

Create method string_to_wsc_oper_class which transforms
strings "24g", "5g" or "24g-5g" to the uint8_t list
with operating class values according to the table
"802.11 table E-4 global operating classes".
Create method string_to_wsc_bss_type which transforms
strings "fronthaul", "backhaul", "fronthaul-backhaul"
to the WSC::eWscVendorExtSubelementBssType.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

MuteSpirit commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -832,3 +832,34 @@ beerocks::eBssType wireless_utils::wsc_to_bwl_bss_type(WSC::eWscVendorExtSubelem
+
+std::list<uint8_t> string_to_wsc_oper_class(const std::string &operating_class)
+{
+    std::list<uint8_t> radio_24g = {81, 82, 83, 84};
+    std::list<uint8_t> radio_5g  = {115, 116, 117, 118, 119, 120, 121, 122,
+                                   123, 124, 125, 126, 127, 128, 129, 130};
```
I think that you may do operating classes lists a global constants instead of local ones:
```
    const std::list<uint8_t> radio_24g{81, 82, 83, 84};
    const std::list<uint8_t> radio_5g{115, 116, 117, 118, 119, 120, 121, 122,
                                                        123, 124, 125, 126, 127, 128, 129, 130};
    const std::list<uint8_t> radio_24g_5g(std::list<uint8_t>(radio_24g).merge(radio_5g))
```

VladyslavTupikin replied:
@MuteSpirit Proposed code does not work.

Commit:
beerocks: tlvf: Fix name for set_wifi_credential message

Current implementation has wrong name for set_wifi_credential messages:
ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST,
ACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE,
cACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST,
cACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE. Also for the new features for
wifi credentials need to change values for the messages in the enum.

Change message names from UPDATE to SET, change values for the request and
response in the enum from 90,91 to 110,111.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
beerocks: tlvf: Add new bml message for update wifi credentials

Before implementation new feature for updating wifi credentials
need to add new messages to the yaml file:
ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST,
ACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE,
cACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST,
cACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE.

Add new messages and set enum values for request,response - 112,113.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
beerocks: tlvf: Add new message type for clear wifi credentials

For the removing wifi credentials for specific AL-MAC need to implement
new messages for yaml file:
ACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST,
ACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE,
cACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST,
cACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE.

Add new message type and assign valuse for request,response
114,115 in the enum.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
controller: bml: Add implementation for bml_clear_wifi_credentials

For removing wifi credentials for specific AL-MAC need to implement
new beerocks_cli command according to the UML diagram
clear_credentials.puml from documents/beerocks_cli.

Implement API according to the uml diagram with all necessary changes:
1. beerocks_cli_bml.cpp         : implement few functions for the
                                  new command for clear credentials
2. bml.cpp                      : implement bml_clear_wifi_credentials
                                  function
3. bml_internal.cpp             : add new function for working with
                                  son_management
4. son_management.cpp           : Add message handler wich removes
                                  wifi crenentials for given AL-MAC

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
controller: bml: Implement bml_update_wifi_credentials

As a part of the new features need to implement new
beerocks_cli command for updating wifi credentials for
the given AL-MAC.

Add implementation for bml_update_wifi_credentials function.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

arnout commented on documentation/images/plantuml/beerocks_cli/set_credentials.puml:
```diff
@@ -0,0 +1,9 @@
+beerocks_cli -> bml_internal: set_wifi_credentials(\n const std::string ssid,\n const std::string pass,\n int sec,\n int vap_id,\n int force)
+bml_internal -> controller: message_com::send_cmdu(m_sockMaster, ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST)
+note right: // TODO: trigger auto-config, probaly\n it will be a good idea to change the message\n to "ACTION_BML_UPDATE_CONFIGURATION_REQUEST"
+@enduml
```
Please make sure there's a newline at the end of the file (nitpick).

VladyslavTupikin replied:
Newline presents in the file.

tomereli replied:
Looks like there still missing a new line (nitpick)

arnout commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,77 @@
+# Introduction
+`bml_set_wifi_credentials` is an old implementation for a CLI command for
+changing credentials for a specific radio. It needs to be revived to allow
+setting credentials for any agent in the network.
```
Please format markdown files according to [the coding style](https://github.com/prplfoundation/prplMesh/blob/7f30baf146b36558379af8b62345e003488c9411/CODINGSTYLE.md#documentation):

> Lines in markdown documents are not wrapped, but every sentence should start on a new line.
> Wrapping in text files is annoying, because adding a single word can cause the entire paragraph to be rewrapped.
> On the other hand, putting everything on one line is not readable either.
> As a compromise, no fixed wrapping is done.
> Instead, each sentence should be started on a new line.
> This way, the diff will only show the sentences that are really affected by the change.
> For very long sentences, subsentences may also be placed on a separate line.

Also (nitpick) for stuff that is in the repository, it is not relevant to say what the old implementation was. So this paragraph can actually be removed.

arnout commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,77 @@
+is configured per alid. Supported parameters are ssid, password (network key),
+encryption type, authentication type, operating class and whether to configure
+as bAP or fAP. For more information, refer to the WSC M2 message defined in the
+WSC specification.
```
This is the important bit.

arnout commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,77 @@
+
+Example for 24g-5g:
+`bml_set_wifi_credentials 11:22:33:44:55:66 multiap maprocks 24g-5g fronthaul-backhaul`
+
```
Please include the png of the flow here:
```
![UML for set_wifi_credentials flow](../..//plantuml/beerocks_cli/clear_credentials.png)
```

(I'm not sure if the path should be relative to this file or not, you'll have to play around a bit after pushing to the branch and doing "view file" to see the result.)

arnout commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,77 @@
+operating_class - can be 24g,5g or 24g-5g, list of operating classes.
+                  Default - 24g-5g[1]
+bss_type - can be fronthaul, backhaul, fronthaul-backhaul.
+           Default - fronthaul.
```
This doesn't render nicely - see https://github.com/prplfoundation/prplMesh/blob/3cd97ed8d6757b771ab20663d5fcf8785471e177/documentation/flows/beerocks_cli/wifi_credentials.md

Better:
```suggestion
- AL-MAC - agent mac address
- ssid - service set identifier
- network_key - password.
              Default - empty password, without security.
- operating_class - can be 24g,5g or 24g-5g, list of operating classes.
                  Default - 24g-5g[1]
- bss_type - can be fronthaul, backhaul, fronthaul-backhaul.
           Default - fronthaul.
```

Renders as:

- AL-MAC - agent mac address
- ssid - service set identifier
- network_key - password.
              Default - empty password, without security.
- operating_class - can be 24g,5g or 24g-5g, list of operating classes.
                  Default - 24g-5g[1]
- bss_type - can be fronthaul, backhaul, fronthaul-backhaul.
           Default - fronthaul.

arnout commented on documentation/images/plantuml/beerocks_cli/clear_credentials.puml:
```diff
@@ -0,0 +1,11 @@
+controller -> controller_db: clear_bss_info_configuration(\n const sMacAddr &al_mac)
+controller -> bml_internal: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE)
+@enduml
+
```
(nitpick) there shouldn't be an empty line at the end of the file either. Just a single newline.

arnout commented on documentation/images/plantuml/beerocks_cli/set_credentials.puml:
```diff
@@ -1,9 +1,11 @@
+controller -> controller_db: add_bss_info_configuration(\n const sMacAddr &al_mac,\n const wireless_utils::sBssInfoConf &bss_info)
+controller -> bml_internal: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_ADD_RESPONSE)
+@enduml
+
```
Single newline.

arnout commented on documentation/images/plantuml/beerocks_cli/update_credentials.puml:
```diff
@@ -0,0 +1,11 @@
+controller -> controller: send_cmdu_to_bus(\n AP_AUTOCONFIGURATION_RENEW_MESSAGE,\n al_mac, db.get_local_bridge_mac());
+controller -> bml_internal: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE)
+@enduml
+
```
Single newline.

arnout commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,77 @@
+
+The BML API should be refactored to allow setting credentials for a specific
+alid, with multiple operating classes.
+
```
There should be a paragraph here that explains the overall flow (from the user point of view):

Since the bml_set_wifi_credentials command only configures a single SSID, and it should be possible to configure multpile SSIDs atomically, the overall flow is as follows:
- clear the current credentials;
- set the new credentials (multiple times for multiple SSIDs);
- update (which triggers the update in the agent atomically).

VladyslavTupikin replied:
fixed

arnout commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,77 @@
+
+* set credentials for alid
+* clear credentials
+* update credentials
```
Please change the order so clear is first, because that's how it is used in practice. Same below.

arnout commented on framework/tlvf/yaml/tlvf/WSC/eWscVendorExt.yaml:
```diff
@@ -5,9 +5,11 @@ _namespace: WSC
   BACKHAUL_BSS: 0x40
   FRONTHAUL_BSS: 0x20
   TEARDOWN: 0x10
+  WRONG_TYPE: 0
```
This is not mentioned in the commit message.

VladyslavTupikin replied:
Fixed commit message.

arnout commented on common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:
```diff
@@ -154,6 +154,8 @@ class wireless_utils {
     static std::string wsc_to_bwl_authentication(WSC::eWscAuth authtype);
     static std::string wsc_to_bwl_encryption(WSC::eWscEncr enctype);
     static beerocks::eBssType wsc_to_bwl_bss_type(WSC::eWscVendorExtSubelementBssType bss_type);
+    static std::list<uint8_t> string_to_wsc_oper_class(const std::string &operating_class);
```
```suggestion
    static std::vector<uint8_t> string_to_wsc_oper_class(const std::string &operating_class);
```
[SL.con.2: Prefer using STL vector by default unless you have a reason to use a different container](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon2-prefer-using-stl-vector-by-default-unless-you-have-a-reason-to-use-a-different-container)

In this case it could even be an array, but it doesn't matter much.

VladyslavTupikin replied:
Decided to keep it as is because changing operating class from the list to vector can cause a lot of changes and it is not a part of this issue, i propose to create issue like refactoring - sBssInfoConf structure, where the operating class list will be changed on the vector.

arnout commented on common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:
```diff
@@ -154,6 +154,8 @@ class wireless_utils {
     static std::string wsc_to_bwl_encryption(WSC::eWscEncr enctype);
     static beerocks::eBssType wsc_to_bwl_bss_type(WSC::eWscVendorExtSubelementBssType bss_type);
+    static std::list<uint8_t> string_to_wsc_oper_class(const std::string &operating_class);
+    static WSC::eWscVendorExtSubelementBssType string_to_wsc_bss_type(const std::string &bss_type);
```
Normally the two functions should have been added in separate commits, but it doesn't matter much so don't bother changing that.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -832,3 +832,34 @@ beerocks::eBssType wireless_utils::wsc_to_bwl_bss_type(WSC::eWscVendorExtSubelem
+
+std::list<uint8_t> string_to_wsc_oper_class(const std::string &operating_class)
+{
+    std::list<uint8_t> radio_24g = {81, 82, 83, 84};
```
```suggestion
    const std::vector<uint8_t> radio_24g {81, 82, 83, 84};
```
- [Prefer `{}` initializer](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-list)
- [SL.con.2: Prefer using STL vector by default unless you have a reason to use a different container](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon2-prefer-using-stl-vector-by-default-unless-you-have-a-reason-to-use-a-different-container)
- [Con.4: Use const to define objects with values that do not change after construction](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rconst-const)

(I think in C++20 we could even use constexpr for vectors, but not in C++14)

VladyslavTupikin replied:
We can't use const because merge() method does not work for const lists.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml:
```diff
@@ -441,3 +441,11 @@ cACTION_BML_CHANNEL_SCAN_DUMP_RESULTS_REQUEST:
+
+cACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST:
+  _type: class
+  params: sMacAddr
```
Please use a better name than just `params`. E.g. `al_mac`.

VladyslavTupikin replied:
Changed params to the al-mac

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml:
```diff
@@ -449,3 +449,11 @@ cACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST:
+
+cACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST:
+  _type: class
+  params: sMacAddr
```
```suggestion
  al_mac: sMacAddr
```

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2094,6 +2108,71 @@ int bml_internal::set_wifi_credentials(const std::string ssid, const std::string
+        }
+    }
+
+    auto config = message_com::create_vs_message<
```
`config` is a weird name for this. `clear_request` maybe?

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2094,6 +2108,71 @@ int bml_internal::set_wifi_credentials(const std::string ssid, const std::string
+        return (-BML_RET_OP_NOT_SUPPORTED);
+    }
+
+    // Initialize the promise for receiving the response
+    m_prmWiFiCredentialsClear = &prmWiFiCredentialsClear;
```
(nitpick) if we set this later (just before sending the message), then we don't need to clear it again in the error paths.

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2094,6 +2108,71 @@ int bml_internal::set_wifi_credentials(const std::string ssid, const std::string
+
+    if (m_sockPlatform == nullptr && !connect_to_platform()) {
+        return (-BML_RET_CONNECT_FAIL);
+    }
```
I don't think we use this?

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2094,6 +2108,71 @@ int bml_internal::set_wifi_credentials(const std::string ssid, const std::string
+    if (!is_local_master()) {
+        LOG(ERROR) << "Command supported only on local master!";
+        // Clear the promise holder
+        m_prmWiFiCredentialsClear = nullptr;
```
m_prmWiFiCredentialsClear wasn't set yet so doesn't need to be cleared.

arnout commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1535,6 +1535,51 @@ void son_management::handle_bml_message(Socket *sd,
+
+        auto al_mac = network_utils::mac_to_string(request->params());
+
+        auto agent_mac = database.get_node_parent_ire(al_mac);
```
That doesn't sound right... the al_mac should be the al_mac already, so no database lookup should be needed...

VladyslavTupikin replied:
Here we are checking database - if al-mac presents in the database we send AP_CONFIGURATION_RENEW_MESSAGE else print error message in the log.

tomereli replied:
I just reviewed the code and indeed `get_node_parent_ire` should return the IRE mac address - in this case it should return the bridge mac, which should be equal to the al_mac.

I think we should instead use the `get_all_connected_ires` API:
```
auto ires = database.get_all_connected_ires();
if (ires.find(request->al_mac() != ires.end()) {
  // send renew
} else {
  // print error
}
```

arnout commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1535,6 +1535,51 @@ void son_management::handle_bml_message(Socket *sd,
+        if (!agent_mac.empty()) {
+            auto cmdu_header =
+                cmdu_tx.create(0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_RENEW_MESSAGE);
+
```
The RENEW message requires 3 TLVs:
- AL-MAC of the controller;
- SupportedRole (= controller);
- SupportedFreqBand.

So you have to add these.

For the freqband, you can include both for now, with a TODO to only include the bands that changed.

arnout reviewed (changes requested):
Excellent commit messages! The commits could be split up a little more, but it's good enough as it is.

I'll want to re-review after my comments have been addressed (even though they're pretty simple to fix), so changes requested.

tomereli commented on framework/tlvf/yaml/tlvf/WSC/eWscVendorExt.yaml:
```diff
@@ -5,9 +5,11 @@ _namespace: WSC
   BACKHAUL_BSS: 0x40
   FRONTHAUL_BSS: 0x20
   TEARDOWN: 0x10
+  WRONG_TYPE: 0
```
IMO we don't need to define a WRONG_TYPE..

VladyslavTupikin replied:
I can rename it as an Invalid type.  We need it for writing new APIs. @arnout What do you think?

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2094,6 +2108,71 @@ int bml_internal::set_wifi_credentials(const std::string ssid, const std::string
+    auto config = message_com::create_vs_message<
+        beerocks_message::cACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST>(cmdu_tx);
+
+    if (config == nullptr) {
```
```suggestion
    if (!config) {
```

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2094,6 +2108,71 @@ int bml_internal::set_wifi_credentials(const std::string ssid, const std::string
+        return (-BML_RET_OP_FAILED);
+    }
+
+    config->params() = al_mac;
```
```suggestion
    config->al_mac() = al_mac;
```

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1535,6 +1535,51 @@ void son_management::handle_bml_message(Socket *sd,
+        auto request =
+            beerocks_header
+                ->addClass<beerocks_message::cACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST>();
+        if (request == nullptr) {
```
```suggestion
        if (!request) {
```

LiorAmram commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1509,6 +1509,32 @@ void son_management::handle_bml_message(Socket *sd,
+        auto response = message_com::create_vs_message<
+            beerocks_message::cACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE>(cmdu_tx,
+                                                                           beerocks_header->id());
+        if (response == nullptr) {
```
```suggestion
        if (!response) {
```

LiorAmram commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1509,6 +1509,32 @@ void son_management::handle_bml_message(Socket *sd,
+        if (response == nullptr) {
+            LOG(ERROR) << "Failed building message cACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE ! ";
+        } else {
+            if (message_com::send_cmdu(sd, cmdu_tx) == false) {
```
```suggestion
            if (!message_com::send_cmdu(sd, cmdu_tx)) {
```

LiorAmram commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1535,6 +1535,51 @@ void son_management::handle_bml_message(Socket *sd,
+        auto response = message_com::create_vs_message<
+            beerocks_message::cACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE>(cmdu_tx,
+                                                                            beerocks_header->id());
+        if (response == nullptr) {
```
```suggestion
        if (!response) {
```

LiorAmram commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1535,6 +1535,51 @@ void son_management::handle_bml_message(Socket *sd,
+            LOG(ERROR) << "Failed building message cACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE ! ";
+        } else {
+            response->error_code() = ret;
+            if (message_com::send_cmdu(sd, cmdu_tx) == false) {
```
```suggestion
            if (!message_com::send_cmdu(sd, cmdu_tx)) {
```

Commit:
controller: master: Implement send_ap_conf_renew_msg

As a part of the new feature for updating wifi credentials
need to implement new method in son_action for sending
AP_CONFIGURATION_RENEW_MESSGE for the given AL-MAC.

Add implementation for the send_ap_conf_renew_msg method.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

arnout commented on controller/src/beerocks/master/son_actions.cpp:
```diff
@@ -479,3 +479,60 @@ bool son_actions::send_cmdu_to_agent(const std::string &dest_mac,
+    auto al_mac_addr    = network_utils::mac_to_string(al_mac);
+    auto controller_mac = database.get_local_bridge_mac();
+
+    auto ires = database.get_all_connected_ires();
+
+    if (ires.find(al_mac_addr) != ires.end()) {
```
:+1: :exclamation:

tomereli commented on documentation/images/plantuml/beerocks_cli/clear_credentials.puml:
```diff
@@ -0,0 +1,10 @@
+bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST)
+controller -> controller_db: clear_bss_info_configuration(\n const sMacAddr &al_mac)
+controller -> bml_internal: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_CLEAR_RESPONSE)
+@enduml
```
Missing line break at end of file.

tomereli commented on documentation/images/plantuml/beerocks_cli/update_credentials.puml:
```diff
@@ -0,0 +1,10 @@
+bml_internal -> controller: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST)
+controller -> controller: send_cmdu_to_bus(\n AP_AUTOCONFIGURATION_RENEW_MESSAGE,\n al_mac, db.get_local_bridge_mac());
+controller -> bml_internal: message_com::send_cmdu(\n m_sockMaster,\n ACTION_BML_WIFI_CREDENTIALS_UPDATE_RESPONSE)
+@enduml
```
Missing line break

tomereli commented on documentation/flows/beerocks_cli/wifi_credentials.md:
```diff
@@ -0,0 +1,74 @@
+![UML for set_wifi_credentials flow](../../images/plantuml/beerocks_cli/update_credentials.png)
+
+# References
+[1] 802.11 table E-4 global operating classes
```
I think you should add more references here:
 
[2] [Multi-AP 2.0 Specification](https://www.wi-fi.org/file/multi-ap-specification)
[3] WSC 2.0.6 specification

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/WSC/eWscVendorExt.h:
```diff
@@ -21,9 +21,11 @@ namespace WSC {
 
 enum eWscVendorExtSubelementBssType {
     BACKHAUL_STA = 0x80,
+    FRONTHAUL_BACKHAUL_BSS = 0x60,
```
We don't need this - I changed this in https://github.com/prplfoundation/prplMesh/pull/1168, the bss_type is a bitmask so in order to set fronthaul and backhaul types we can simply set:
```
bss_type = eWscVendorExtSubelementBssType::BACKHAUL_BSS | eWscVendorExtSubelementBssType::FRONTHAUL_BSS;
```

tomereli replied:
Relevant commits:
https://github.com/prplfoundation/prplMesh/pull/1168/commits/02bad4695b78f613a84bc8fa79818bd781f801ae
https://github.com/prplfoundation/prplMesh/pull/1168/commits/2dcf4ba4d1c49409494ab9ab0a59b493e2654c39

VladyslavTupikin replied:
fixed

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/WSC/eWscVendorExt.h:
```diff
@@ -21,9 +21,11 @@ namespace WSC {
     BACKHAUL_BSS = 0x40,
     FRONTHAUL_BSS = 0x20,
     TEARDOWN = 0x10,
+    ERROR_TYPE = 0x0,
```
I don't understand why we need an error type.

VladyslavTupikin replied:
fixed

tomereli commented on framework/tlvf/yaml/tlvf/WSC/eWscVendorExt.yaml:
```diff
@@ -5,9 +5,11 @@ _namespace: WSC
 eWscVendorExtSubelementBssType:
   _type: enum
   BACKHAUL_STA: 0x80
+  FRONTHAUL_BACKHAUL_BSS: 0x60 # (FRONTHAUL_BSS | BACKHAUL_BSS)
```
Same here, I think we should remove this commit completely. We don't need to define new values in bitmask enums.

VladyslavTupikin replied:
fixed

tomereli commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -832,3 +832,35 @@ beerocks::eBssType wireless_utils::wsc_to_bwl_bss_type(WSC::eWscVendorExtSubelem
+}
+
+WSC::eWscVendorExtSubelementBssType
+wireless_utils::string_to_wsc_bss_type(const std::string &bss_type)
```
```suggestion
uint8_t wireless_utils::string_to_wsc_bss_type(const std::string &bss_type)
{
    if (bss_type == "backhaul") {
        return WSC::eWscVendorExtSubelementBssType::BACKHAUL_BSS;
    } else if (bss_type == "fronthaul") {
        return WSC::eWscVendorExtSubelementBssType::FRONTHAUL_BSS;
    } else if (bss_type == "fronthaul-backhaul") {
        return WSC::eWscVendorExtSubelementBssType::FRONTHAUL_BSS | WSC::eWscVendorExtSubelementBssType::BACKHAUL_BSS;
    }
    
    return 0; // error
}
```

VladyslavTupikin replied:
fixed

tomereli commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h:
```diff
@@ -356,8 +356,6 @@ enum eActionOp_BML: uint8_t {
     ACTION_BML_SET_DFS_REENTRY_RESPONSE = 0x55,
     ACTION_BML_GET_DFS_REENTRY_REQUEST = 0x56,
     ACTION_BML_GET_DFS_REENTRY_RESPONSE = 0x57,
-    ACTION_BML_WIFI_CREDENTIALS_UPDATE_REQUEST = 0x5a,
```
I prefer the autogenerated changes in a separate commit (nitpick)

VladyslavTupikin replied:
@tomereli I think keeping autogenerated files in the repository bad idea.

arnout committed:
test_flows.py: add ap_config_bss_tear_down_cli test

This is the same test as ap_config_bss_tear_down, but using beerocks_cli
commands instead of CAPI commands. Its main purpose is to test
beerocks_cli.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1535,6 +1535,35 @@ void son_management::handle_bml_message(Socket *sd,
+        }
+
+        ret = son_actions::send_ap_config_renew_msg(cmdu_tx, database, request->al_mac());
+        if (!ret) {
```
```suggestion
       LOG_IF(!ret, ERROR) << "Failed son_actions::send_ap_config_renew_msg!";
```

VladyslavTupikin replied:
I prefer to keep as is. The current string is compatible with the whole of the file and I do not see any reason to change it on the proposed macro. If you want to start using it - need to create a separate issue for refactoring code in this file.

tomereli commented on controller/src/beerocks/master/son_actions.cpp:
```diff
@@ -479,3 +479,60 @@ bool son_actions::send_cmdu_to_agent(const std::string &dest_mac,
+    }
+
+    return result;
+}
```
Missing line break at end of file.

tomereli commented on tests/test_flows.py:
```diff
@@ -167,6 +167,51 @@ def test_ap_config_bss_tear_down(self):
+        # Same test as the previous one but using CLI instead of dev_send_1905
+
+        env.beerocks_cli_command('bml_clear_wifi_credentials {}'.format(env.agents[0].mac))
+        # TODO replace with bml_set_wifi_credentials
```
I don't understand than why add this test now if you can't use the new APIs yet. Should this commit perhaps belong to the next PR?

VladyslavTupikin replied:
@arnout Please explain your point of view

arnout replied:
@tomereli should learn to read :-P The lines above and below test the new commands introduced by this PR (clear and update).

tomereli replied:
My bad.

tomereli reviewed (commented):
Nothing major, but please take my comments into consideration before merging, especially the one about the FRONTHAUL_BACKHAUL and ERROR enum which I think should be dropped.
